nemo/collections/nlp/models/intent_slot_classification/intent_slot_classification_model.py

```diff
@@ -53,70 +53,124 @@ def output_types(self) -> Optional[Dict[str, NeuralType]]:
-            self.intent_loss = CrossEntropyLoss(logits_ndim=2, weight=self.data_desc.intent_weights)
-            self.slot_loss = CrossEntropyLoss(logits_ndim=3, weight=self.data_desc.slot_weights)
+            self.intent_loss = CrossEntropyLoss(logits_ndim=2, weight=self.cfg.intent_weights)
+            self.slot_loss = CrossEntropyLoss(logits_ndim=3, weight=self.cfg.slot_weights)
         else:
             self.intent_loss = CrossEntropyLoss(logits_ndim=2)
             self.slot_loss = CrossEntropyLoss(logits_ndim=3)
 
-        self.total_loss = AggregatorLoss(num_inputs=2, weights=[cfg.intent_loss_weight, 1.0 - cfg.intent_loss_weight])
+        self.total_loss = AggregatorLoss(
+            num_inputs=2, weights=[self.cfg.intent_loss_weight, 1.0 - self.cfg.intent_loss_weight]
+        )
 
         # setup to track metrics
         self.intent_classification_report = ClassificationReport(
-            num_classes=self.data_desc.num_intents,
-            label_ids=self.data_desc.intents_label_ids,
+            num_classes=len(self.cfg.data_desc.intent_labels),
+            label_ids=self.cfg.data_desc.intent_label_ids,
             dist_sync_on_step=True,
             mode='micro',
         )
         self.slot_classification_report = ClassificationReport(
-            num_classes=self.data_desc.num_slots,
-            label_ids=self.data_desc.slots_label_ids,
+            num_classes=len(self.cfg.data_desc.slot_labels),
+            label_ids=self.cfg.data_desc.slot_label_ids,
             dist_sync_on_step=True,
             mode='micro',
         )
 
-    def update_data_dir(self, data_dir: str) -> None:
+    def update_data_dir_for_training(self, data_dir: str, train_ds, validation_ds) -> None:
         """
         Update data directory and get data stats with Data Descriptor
         Weights are later used to setup loss
 
         Args:
             data_dir: path to data directory
         """
+        logging.info(f'Setting data_dir to {data_dir}.')
+        self.data_dir = data_dir
+        # Update configuration with new data.
+        self._copy_data_desc_to_cfg(self.cfg, data_dir, train_ds, validation_ds)
+        # Finish the "conditional initialization" by passing the new data_dir.
+        self._setup_tokenizer(self.cfg.tokenizer)
+        self._init_modules_when_tokenizer_set()
+
+    def update_data_dir_for_testing(self, data_dir) -> None:
+        """
+        Update data directory.
+        Weights are later used to setup loss
+
+        Args:
+            data_dir: path to data directory
+        """
+        logging.info(f'Setting data_dir to {data_dir}.')
         self.data_dir = data_dir
-        logging.info(f'Setting model.data_dir to {data_dir}.')
+        # Finish the "conditional initialization" by passing the new data_dir.
+        self._setup_tokenizer(self.cfg.tokenizer)
+        self._init_modules_when_tokenizer_set()
```

Comment:
Please consider renaming the method, it's doing way more than changing the data_dir + please update the docstring

Reply:
done